tsz-compress/src/v2/encode.rs

```diff
@@ -0,0 +1,751 @@
+            if (index <= 2 && !(-32768..=32767).contains(&value)) {
+                sixteen = false;
+            }
+            if (index <= 3 && !(-512..=511).contains(&value)) {
+                ten = false;
+            }
+            if (index <= 4 && !(-128..=127).contains(&value)) {
+                eight = false;
+            }
+            if (index <= 8 && !(-8..=7).contains(&value)) {
+                four = false;
+            }
+            if (index <= 10 && !(-4..=3).contains(&value)) {
+                three = false;
+            }
+        });
+
+        // Emit according to priority of cases
+        if three {
+            push_header_pad_three_bits(out);
+            for _ in 0..10 {
+                if let Some(value) = self.pop() {
+                    let value = (value << 1i64) ^ (value >> 63i64); // ZigZag Encoding
+                    value.extend_bits(0..3, out);
+                }
+            }
+            return 10;
+        } else if four {
+            push_header_pad_four_bits(out);
+            for _ in 0..8 {
+                if let Some(value) = self.pop() {
+                    let value = (value << 1i64) ^ (value >> 63i64); // ZigZag Encoding
+                    value.extend_bits(0..4, out);
+                }
+            }
+            return 8;
+        } else if eight {
+            push_header_pad_eight_bits(out);
+            for _ in 0..4 {
+                if let Some(value) = self.pop() {
+                    let value = (value << 1i64) ^ (value >> 63i64); // ZigZag Encoding
+                    value.extend_bits(0..8, out);
+                }
+            }
+            return 4;
+        } else if ten {
+            push_header_pad_ten_bits(out);
+            for _ in 0..3 {
+                if let Some(value) = self.pop() {
+                    let value = (value << 1i64) ^ (value >> 63i64); // ZigZag Encoding
+                    value.extend_bits(0..10, out);
+                }
+            }
+            return 3;
+        } else if sixteen {
+            push_header_pad_sixteen_bits(out);
+            for _ in 0..2 {
+                if let Some(value) = self.pop() {
+                    let value = (value << 1i64) ^ (value >> 63i64); // ZigZag Encoding
+                    value.extend_bits(0..16, out);
+                }
+            }
+            return 2;
+        }
+        0
+    }
+}
+
+impl EmitDeltaBits<i32> for CompressionQueue<i32, 10> {
+    #[allow(unused)]
+    fn emit_delta_bits(&mut self, out: &mut BitBuffer, flush: bool) -> usize {
+        let mut three = true;
+        let mut four = true;
+        let mut eight = true;
+        let mut ten = true;
+        let mut sixteen = true;
+
+        let queue_length = self.len();
+
+        // Check flush conditions
+        if flush {
+            // Can not emit with any case of delta compression if queue is empty or contains 1 sample.
+            if self.is_empty() || self.len() == 1 {
+                return 0;
```

Comment:
On flush, the data has to be emitted. It may require emitting delta-delta format.